inc/class-contributor.php

```diff
@@ -0,0 +1,235 @@
+		}
+		if ( ! in_array( $contributor_type, $this->valid, true ) ) {
+			return '';
+		}
+
+		// Look if contributors exist as taxonomies (new data model)
+		$contributors = [];
+		$meta = get_post_meta( $post_id, $contributor_type, false );
+		foreach ( $meta as $slug ) {
+			$term = get_term_by( 'slug', $slug, 'contributor' );
+			if ( $term ) {
+				$first_name = get_term_meta( $term->term_id, 'contributor_first_name', true );
+				$last_name = get_term_meta( $term->term_id, 'contributor_last_name', true );
+				if ( ! empty( $first_name ) && ! empty( $last_name ) ) {
+					$contributors[] = "{$first_name} {$last_name}";
+				} elseif ( ! empty( $term->name ) ) {
+					$contributors[] = $term->name;
+				}
+			}
+		}
+
+		if ( empty( $contributors ) ) {
+			// Look if contributors exist as metadata (old data model)
+			// If yes then convert to taxonomies (new data model)
+			$map = [
+				'pb_authors' => [ 'pb_author', 'pb_section_author' ],
+				'pb_editors' => [ 'pb_editor' ],
+				'pb_translators' => [ 'pb_translator' ],
+				'pb_contributors' => [ 'pb_contributing_authors' ],
+			];
+			if ( isset( $map[ $contributor_type ] ) ) {
+				foreach ( $map[ $contributor_type ] as $slug ) {
+					$meta = get_post_meta( $post_id, $slug, false );
+					foreach ( $meta as $name ) {
+						$result = $this->insert( $name );
+						if ( $result !== false ) {
+							$added = $this->link( $result['term_id'], $post_id, $contributor_type );
+							if ( $added !== false ) {
+								$contributors[] = $name;
+							}
+						}
+					}
+				}
+			}
+		}
+
+		// Comma separated, Oxford comma, localized and between the last two items
+		if ( count( $contributors ) === 2 ) {
+			return $contributors[0] . ' ' . __( 'and', 'pressbooks' ) . ' ' . $contributors[1];
+		} else {
+			$last = array_pop( $contributors );
+			$output = implode( ', ', $contributors );
+			if ( $output ) {
+				$output .= ', ' . __( 'and', 'pressbooks' ) . ' ';
+			}
+			$output .= $last;
+			return $output;
+		}
+	}
+
+	/**
+	 * @param string $full_name
+	 * @param int $post_id (optional)
+	 * @param string $contributor_type (optional)
+	 *
+	 * @return array|false An array containing the `term_id` and `term_taxonomy_id`, false otherwise.
+	 */
+	public function insert( $full_name, $post_id = 0, $contributor_type = 'pb_authors' ) {
+
+		$full_name = trim( $full_name );
+		$slug = sanitize_title_with_dashes( remove_accents( $full_name ), '', 'save' );
+		$term = get_term_by( 'slug', $slug, 'contributor' );
+		if ( $term ) {
+			return [
+				'term_id' => $term->term_id,
+				'term_taxonomy_id' => $term->term_taxonomy_id,
+			];
+		}
+		$results = wp_insert_term(
+			$full_name, 'contributor', [
+				'slug' => $slug,
+			]
+		);
+
+		if ( $post_id && is_array( $results ) ) {
+			$this->link( $results['term_id'], $post_id, $contributor_type );
+		}
+
+		return is_array( $results ) ? $results : false;
+	}
+
+	/**
+	 * Associate a Contributor's Term ID to a Post ID (Taxonomy + Meta)
```

Comment:
Technically we are assigning the Term Slug to the Post ID? Maybe good to clarify in the code comment?